### ISSUE TITLE
Create analogous solution for replay bloc

### ISSUE THREAD

**Issue by GregoryConrad** [GregoryConrad/rearch-dart#35]
Probably will involve a new side effect for handling a stack of states.

**Comment by GregoryConrad** [comment]
I'm thinking an API like:

```dart
Replay<T> = (
  T?, void Function(T), {
  void Function()? undo,
  void Function()? redo,
});
final (currData, addData, :undo, :redo) = use.replay();
```

Some open questions:
- `use.replay` or `use.stateStack`?
  - Inclination is `use.replay` to show the inspiration
- ~Should the side effect also take a current state as its input as an alternative to `addData`?~
  - Probably not, as how will we know when it is called with null or non-null? We would need to force users to pass in `Option<T>` (since you can't conditionally pass in a parameter to a function in dart yet, the const object default parameter hack wouldn't work as users need to unconditionally call the side effect)

Implementation note: any call to `addData` should delete the rest of the stack (redo-es are invalidated when new data is saved from an earlier state).

**Comment by GregoryConrad** [comment]
Perhaps a composition can be done via:

```dart
final (data, addData, :undo, :redo) = use.replay();
final (persistedState, persistData) = use.persist(/* ... */);
final txn = use.transactionRunner();
final save = use.memo(() {
  return (T data) {
    txn(() {
      addData(data);
      persistData(data);
    });
  };
}, [txn, addData, persistData]);
final undoAndSave = use.memo(() {
  return () {
    txn(() {
      persistData(undo());
    });
  };
}, [txn, undo, persistData]);
final redoAndSave = use.memo(() {
  return () {
    txn(() {
      persistData(redo());
    });
  };
}, [txn, redo, persistData]);
```
After the last line of code:
- `data` will be the optimistic new data
- `persistedState` will be the pessimistic persisted data state
- `save`, `undoAndSave`, and `redoAndSave` will all do exactly as they sound

However, returning `T` from `undo` and `redo` in the first place feels wrong--this can encourage breaking reactivity.

Alternatively, we could do something like this:

```dart
final (persistedState, persistData) = use.persist(/* ... */);
final (data, addData, :undo, :redo) = use.replay(onChange: persistData);
```

Care would have to be taken in the `replay` implementation to make sure we always will call the last-passed-in `onChange` listener.

**Comment by GregoryConrad** [comment]
Ahh! I just realized something I hadn't before. `persist` was designed to be able to control its own state. I actually ran into a similar problem when writing [my blog post](https://blog.gsconrad.com/2023/12/22/the-problem-with-state-management.html), but I didn't come to this realization then and instead had to use a side effect transaction there too (which was a bit out of the scope of the blog post, but necessary given the available side effects at the time).

In the example above, I was trying to force two different side effects to sync their states; this can be difficult. Here's the proper solution: let one side effect control the overarching state and designate other side effects to use that state. Thus, we should probably introduce a new side effect like:

```dart
final (data, addData, :undo, :redo) = use.replay();
final persistState = use.cache(Some(data)); // or persistCurrent?
// If use.cache has None input, it won't attempt to cache anything new
// (so we don't override any existing persisted values).
// If Some, it will attempt to cache the current input 
```

And implement persist as a simple `use.state()` coupled with this new side effect to maintain backwards compatibility. Put a big header in the persist doc that you normally should `use.state()` and `use.cache()` (or whatever we name it) unless you are directly returning the output of `persist`
